http/native_server.ts

```diff
@@ -0,0 +1,358 @@
+      if (
+        // The listener is closed
+        error instanceof Deno.errors.BadResource ||
+        // TLS handshake errors
+        error instanceof Deno.errors.InvalidData ||
+        error instanceof Deno.errors.UnexpectedEof ||
+        error instanceof Deno.errors.ConnectionReset ||
+        error instanceof Deno.errors.NotConnected
+      ) {
+        return mux.add(this.acceptConnAndIterateHttpRequests(mux));
+      }
+
+      throw error;
+    }
+
+    // "Upgrade" the network connection into a HTTP connection
+    const httpConn = Deno.serveHttp(conn);
+
+    // Closing the underlying server will not close the HTTP connection,
+    // so we track it for closure upon shutdown.
+    this.trackConnection(httpConn);
+
+    // Try to accept another connection and add it to the multiplexer.
+    mux.add(this.acceptConnAndIterateHttpRequests(mux));
+
+    // Yield the requests that arrive on the just-accepted connection.
+    yield* this.iterateHttpRequests(httpConn, conn);
+  }
+
+  /**
+   * Adds the HTTP connection to the internal tracking list.
+   *
+   * @param {Deno.HttpConn} httpConn
+   * @private
+   */
+  private trackConnection(httpConn: Deno.HttpConn): void {
+    this.#httpConnections.add(httpConn);
+  }
+
+  /**
+   * Removes the HTTP connection from the internal tracking list.
+   *
+   * @param {Deno.HttpConn} httpConn
+   * @private
+   */
+  private untrackConnection(httpConn: Deno.HttpConn): void {
+    this.#httpConnections.delete(httpConn);
+  }
+
+  /**
+   * Implementation of Async Iterator to allow consumers to loop over
+   * HTTP requests.
+   *
+   * @returns {AsyncIterableIterator<ServerRequest>} The async iterator.
+   */
+  [Symbol.asyncIterator](): AsyncIterableIterator<ServerRequest> {
+    const mux: MuxAsyncIterator<ServerRequest> = new MuxAsyncIterator();
+    mux.add(this.acceptConnAndIterateHttpRequests(mux));
+
+    return mux.iterate();
+  }
+}
+
+/**
+ * Accept incoming connections on the given listener, and handle requests on
+ * these connections with the given handler.
+ *
+ *     const listener = Deno.listen({ port: 8000 });
+ *
+ *     serve(listener, (request) => {
+ *       const body = `Your user-agent is:\n\n${request.headers.get(
+ *         "user-agent",
+ *       ) ?? "Unknown"}`;
+ *
+ *       return new Response(body, { status: 200 });
+ *     });
+ *
+ * @param {Deno.Listener} listener The listener to accept connections from.
+ * @param {HTTPHandler} handler The handler for individual HTTP requests.
+ * @param {ServerInit} [init] Additional server options.
+ */
+export async function serve(
+  listener: Deno.Listener,
+  handler: HTTPHandler,
+  init: ServerInit = {},
+): Promise<void> {
+  const server = new Server(listener);
+
+  if (init?.signal) {
+    init.signal.onabort = () => {
+      server.close();
+    };
+  }
+
+  for await (const requestEvent of server) {
+    requestEvent.respondWith(
+      handler(requestEvent.request, requestEvent.connInfo),
+    );
```

Comment:
Main open question I have is are we ok with this potentially throwing and not catching (so iterator loop ends etc.)?

This is same for the existing server API implementations (REF: https://github.com/denoland/deno_std/blob/main/http/server.ts#L327) so perhaps is ok under the assumption that anyone wanting to have fuller control over their request loop would construct a server themselves

```ts
import { Server } from "https://deno.land/std@$STD_VERSION/http/native_server.ts";

// ... rest of code

const server = new Server(listener);

for await (const requestEvent of server) {
  // ... consumer handles the requestEvent with additional error handling etc.
}
```

Given the RFC proposal of an http handler passed in I don't think we have much of a choice but to be opionated on behalf of the consumer. Alternatives include us `.catch(() => {})` swallowing all errors, or perhaps extending the `ServerInit` or sim. to accept an `onRequestError(error: Error): Promise<void>`. Perhaps other options as well?

Reply:
If following precedent of https://pkg.go.dev/net/http#Handler then perhaps we should recover unless the error is of a certain instance